Fall back on `sys.real_prefix` for Python 2.7

dandersson opened the pull request:
`sys.base_prefix` is used in the SQLite3 import hook from pull request #261. This attribute was introduced in Python 3.3, which makes this hook raise an uncaught exception when running Python 2.7.

Fall back to looking at `sys.real_prefix`, which is set by `virtualenv` to the equivalent of `sys.base_prefix` when running in a virtual environment, or just `sys.prefix` otherwise.

Since this regards Python 2.7 compatibility, this should only be relevant for the cx_Freeze 5.x branch.

Sample traceback from the kind of error this intends to fix:

```
running build
running build_exe
Traceback (most recent call last):
  File "config/cx_freeze_setup.py", line 315, in <module>
    main()
  File "config/cx_freeze_setup.py", line 287, in main
    executables=executables
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\dist.py", line 349, in setup
    distutils.core.setup(**attrs)
  File "C:\python27\Lib\distutils\core.py", line 151, in setup
    dist.run_commands()
  File "C:\python27\Lib\distutils\dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "C:\python27\Lib\distutils\dist.py", line 972, in run_command
    cmd_obj.run()
  File "C:\python27\Lib\distutils\command\build.py", line 127, in run
    self.run_command(cmd_name)
  File "C:\python27\Lib\distutils\cmd.py", line 326, in run_command
    self.distribution.run_command(command)
  File "C:\python27\Lib\distutils\dist.py", line 972, in run_command
    cmd_obj.run()
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\dist.py", line 219, in run
    freezer.Freeze()
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\freezer.py", line 621, in Freeze
    self.finder = self._GetModuleFinder()
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\freezer.py", line 340, in _GetModuleFinder
    finder.IncludePackage(name)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 655, in IncludePackage
    self._ImportAllSubModules(module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 287, in _ImportAllSubModules
    recursive)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 279, in _ImportAllSubModules
    deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 569, in _ScanCode
    fromList, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 210, in _EnsureFromList
    self._ImportModule(subModuleName, deferredImports, caller)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 310, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 564, in _ScanCode
    module, relativeImportIndex)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 310, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 588, in _ScanCode
    topLevel = False)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 564, in _ScanCode
    module, relativeImportIndex)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 310, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 416, in _LoadModule
    namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 485, in _LoadPackage
    self._LoadModule(name, fp, path, info, deferredImports, parent)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 564, in _ScanCode
    module, relativeImportIndex)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 310, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 588, in _ScanCode
    topLevel = False)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 588, in _ScanCode
    topLevel = False)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 564, in _ScanCode
    module, relativeImportIndex)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 310, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 384, in _InternalImportModule
    namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 416, in _LoadModule
    namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 485, in _LoadPackage
    self._LoadModule(name, fp, path, info, deferredImports, parent)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 588, in _ScanCode
    topLevel = False)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 564, in _ScanCode
    module, relativeImportIndex)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 310, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 564, in _ScanCode
    module, relativeImportIndex)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 310, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 564, in _ScanCode
    module, relativeImportIndex)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 310, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 564, in _ScanCode
    module, relativeImportIndex)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 310, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 416, in _LoadModule
    namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 485, in _LoadPackage
    self._LoadModule(name, fp, path, info, deferredImports, parent)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 564, in _ScanCode
    module, relativeImportIndex)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 345, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 564, in _ScanCode
    module, relativeImportIndex)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 326, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 416, in _LoadModule
    namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 485, in _LoadPackage
    self._LoadModule(name, fp, path, info, deferredImports, parent)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 588, in _ScanCode
    topLevel = False)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 569, in _ScanCode
    fromList, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 210, in _EnsureFromList
    self._ImportModule(subModuleName, deferredImports, caller)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 310, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 474, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 588, in _ScanCode
    topLevel = False)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 588, in _ScanCode
    topLevel = False)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 564, in _ScanCode
    module, relativeImportIndex)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 310, in _ImportModule
    deferredImports, namespace = namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 403, in _InternalImportModule
    parentModule, namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 416, in _LoadModule
    namespace)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 485, in _LoadPackage
    self._LoadModule(name, fp, path, info, deferredImports, parent)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 463, in _LoadModule
    self._RunHook("load", module.name, module)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\finder.py", line 536, in _RunHook
    method(self, *args)
  File "C:\dev\test_sqlite3_hook\env\lib\site-packages\cx_Freeze\hooks.py", line 826, in load_sqlite3
    dll_path = os.path.join(sys.base_prefix, "DLLs", dll_name)
AttributeError: 'module' object has no attribute 'base_prefix'
```

Commit:
Fall back on `sys.real_prefix` for Python 2.7

`sys.base_prefix` was not introduced until Python 3.3 which made this
hook raise an uncaught exception in Python 2.7. Catch this and fall back
to `sys.real_prefix` if it is set (i.e. we are in a virtual environment
-- this is set by `activate_this.py` when activating a virtual
environment from the `virtualenv` package) or `sys.prefix` otherwise.

Since this regards Python 2.7 compatibility, this should only be
relevant for the cx_Freeze 5.x branch.

FreeHarry commented:
@dandersson: Thanks for improving my pull request #261! I forgot to check it for older python versions.

anthony-tuininga commented on cx_Freeze/hooks.py:
```diff
@@ -816,12 +816,23 @@ def load_clr(finder, module):
+    except AttributeError:
+        # For Python 2, use sys.real_prefix set by virtualenv if set,
+        # otherwise fall back to sys.prefix
+        return getattr(sys, "real_prefix", "prefix")
```
This code doesn't do what I think you want it to do -- namely, use sys.real_prefix unless it cannot be found, in which case return sys.prefix. Perhaps something like this instead?

return getattr(sys, "real_prefix", sys.prefix)

dandersson replied:
Absolutely correct -- I will update the pull request.

dandersson committed:
Fix fallback to sys.prefix

Fix usage of `getattr`, which previously would have returned the literal
string "prefix" as a fallback for `sys.real_prefix` instead of
`sys.prefix` when running cx_Freeze from a global Python 2.7
installation, as opposed to from within a virtual environment.

wlach commented:
@anthony-tuininga would it be possible to release a new version of the library in the 5.x series with this fix? I'd like to use a later version with mozregression, which is stuck on 2.7 for now.

anthony-tuininga commented:
Yes. Will release a new version of cx_Freeze 5.x soon. Its overdue. :-)

lcunha0118 commented:
@anthony-tuininga I have version 5.0 and I still have the same error. And the hooks.py file was modified so the same fix does not apply. Was the problem fixed?

tnsolley commented:
I can confirm this issue is still present in version 5.02 of cx_Freeze.

mattbk commented:
> I can confirm this issue is still present in version 5.02 of cx_Freeze.

Ditto.

ourish commented:
> I can confirm this issue is still present in version 5.02 of cx_Freeze.

As of 3/11 the issue remains
